src/main/scala/com/gu/contentapi/firehose/kinesis/SingleEventProcessor.scala

```diff
@@ -70,14 +69,27 @@ abstract class EventProcessor[EventT <: ThriftStruct: ThriftStructCodec]
     recordsProcessedSinceCheckpoint.set(0)
   }
 
-  /* This method may be called by KCL, e.g. in case of shard splits/merges */
-  override def shutdown(checkpointer: IRecordProcessorCheckpointer, reason: ShutdownReason): Unit = {
-    if (reason == ShutdownReason.TERMINATE) {
-      checkpointer.checkpoint()
-    }
-    logger.info(s"Shutdown event processor for shard $shardId because $reason")
+  def leaseLost(leaseLostInput: LeaseLostInput): Unit = {
```

Comment:
Do these new methods alter the likelihood of being disconnected from the stream at all? Thinking in terms of e.g. the cache purger, if we're shut down like this will we need more defensive code there?

Reply:
I don't believe so. They have been introduced only because the API of the library changed, but I think that the actual functionality is the same. I'll double-check it all though

Reply:
To expand on that point - KCL v1 expects a single method, `shutdown`, with a nullable `checkpointer` parameter and a `ShutdownReason` parameter provided to it by the library.
`checkpointer` is only valid if it's a requested shutdown.

KCL v2 dispenses with the `ShutdownReason` parameter and instead defines separate overridable methods for each of the valid shutdown reasons.  These, in turn, have their own parameters as the `{EventName}Input` typed parameters.  `ShutdownRequestedInput` includes a _non-nullable_ `checkpointer` reference and the others don't include one at all.

To maintain client compatibility, I've defined an overridable `shutdown` method that does maintain the `ShutdownReason` parameter which can be used by any client needing notification.  As before, this is not mandatory; in KCL v1 the old implementation handled the safe checkpointing and shutting down. In the new implementation, this is done over the 3 new callbacks and they then call the `shutdown` method to allow a client to be notified.

Reply:
Thanks!